Save ensemble/cluster state when cluster enabled

andrewjstone opened the pull request:
Fixes #34 and hopefully basho/riak_kv#983
See #34 for details

andrewjstone committed:
Save ensemble/cluster state when cluster enabled

Fixes #34 and hopefully basho/riak_kv#983
See #34 for details

Note that even when `riak_ensemble_manager:maybe_save_state/1` fails during
enable, we still want to call `riak_ensemble_manager:state_changed/1` to
ensure that we don't have inconsistent ets tables and running peers.
However, since this activation is not saved we can still send up with
the issue in basho/riak_kv#983.

engelsanchez commented on src/riak_ensemble_manager.erl:
```diff
@@ -297,6 +297,7 @@ handle_call(enable, _From, State) ->
     case activate(State) of
         {ok, State2} ->
             State3 = state_changed(State2),
+            ok = riak_ensemble_storage:sync(),
```
@andrewjstone I don't think that either activate/1 nor state_change actually save the state, so the sync will not have the desired effect. Look at the paths that call maybe_save_state, like the clause for {join, Node}.

andrewjstone replied:
Whoops. I was thinking for some reason that sync retrieved that data from ets. Itchy trigger finger today. Will fix.

engelsanchez commented on src/riak_ensemble_manager.erl:
```diff
@@ -312,6 +312,7 @@ handle_call({join, Node}, _From, State=#state{cluster_state=LocalCS}) ->
                     State2 = State#state{cluster_state=RemoteCS},
                     case maybe_save_state(State2) of
                         ok ->
+                            ok = riak_ensemble_storage:sync(),
```
Dude, somebody needs some coffee today :)
This moves the sync to the join case, which I mentioned as an example of a path that saves the state to disk.  I believe the right way to do this is to call maybe_save_state after activating the cluster.  If that fails for any weird reason, it would be safer to reply with error and not go on. That way we don't have to deal with an activated state that wasn't saved to disk.  Then later call state_changed on it so peers are started and finish as before. What do you think?

andrewjstone replied:
Damnit. I realized that. I conflated joining and enabling. Your suggestion is correct. I'm going for a walk. Will fix this when I get back and am more awake and less distracted.

engelsanchez commented:
From the commit

```
Note that even when `riak_ensemble_manager:maybe_save_state/1` fails during
enable, we still want to call `riak_ensemble_manager:state_changed/1` to
ensure that we don't have inconsistent ets tables and running peers.
However, since this activation is not saved we can still send up with
the issue in basho/riak_kv#983.
```

I think the best way to handle a failure to save is to avoid changing the state at all. That is, return an error code and the original state. Since activate doesn't have any side effects, this should be safe. The state will remain as it was before and the operation could be retried. Currently, you can not retry an activation (it will just return an error).  I think that would make more sense.

andrewjstone commented:
acitvate is not pure because it calls `ets:insert`, which is why the code looks like it does. After discussing with @engelsanchez we decided to refactor the insert outside of that activate and not change state if save fails. That will be forthcoming in another commit.

andrewjstone committed:
Ensure activated cluster state is saved before use

In order to do this we turn `riak_ensemble_manager:activate/0` into a
pure function. We do this by removing calls to `ets:insert/2` and
returning the data to be inserted instead. The caller can then attempt
to save and sync the data. If this succeeds, then finally the call to
`ets:insert/2` is made and the state is updated along with `ok` being
returned to the client.

engelsanchez commented:
+1 cc74c78
Latest code looks good and all ensemble tests are passing with it

andrewjstone commented:
@borshop merge
